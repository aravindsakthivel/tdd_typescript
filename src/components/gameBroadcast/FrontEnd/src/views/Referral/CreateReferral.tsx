import React, { useState, ReactElement } from 'react';
import { useMutation } from '@apollo/client';
import { Select, Button, Form } from 'antd';
import ReferralRowTemplate from './ReferralRowTemplate';
import ReferralLinkModal from './ReferralLinkModal';
import { SuccessMessage, ErrorMessage } from '../Common/SequentialMessage';
import { GENERATE_REFERRAL_LINK } from '../../queries';
import {
  SelectOptionInterface,
  DestinationInterface,
} from './ReferralInterfaces';

const { Option } = Select;

function CreateReferral({
  destination: destinationTest = '',
}: DestinationInterface): ReactElement {
  const [generateReferrals, { data, loading, error }] = useMutation(
    GENERATE_REFERRAL_LINK,
    {
      notifyOnNetworkStatusChange: true,
    },
  );
  const [isModalOpen, setIsModalOpen] = useState<boolean>(false);
  const [destination, setDestination] = useState<string>(destinationTest);
  // currently no destination info present using general values
  const optionValues: SelectOptionInterface[] = [
    {
      name: 'MPL PRO',
      value: 'mpl_pro',
    },
    {
      name: 'MPL QA',
      value: 'mpl_qa',
    },
    {
      name: 'MPL DEV',
      value: 'mpl_dev',
    },
  ];

  const onChangeDestination = (option: string) => {
    setDestination(option);
  };

  const onSubmit = () => {
    generateReferrals({
      variables: {
        generateReferralLinkData: { userId: 'skdfjlksj589', destination },
      },
    })
      .then(() => {
        SuccessMessage('Referral Link Generated Successfully', 3);
      })
      .catch(() => {
        ErrorMessage('Something went wrong try again', 3);
      });
    setIsModalOpen(!isModalOpen);
  };

  return (
    <>
      <ReferralRowTemplate title="Create a custom link" last={false}>
        <Form
          layout="inline"
          size="large"
          onFinish={onSubmit}
          data-testid="generate-link-form"
          initialValues={
            destinationTest !== '' ? { destination: destinationTest } : {}
          }
          className="referral-link-form">
          <Form.Item
            name="destination"
            rules={[
              { required: true, message: 'Please select a Destination' },
            ]}>
            <Select
              style={{ width: 350, borderRadius: 10 }}
              placeholder="Select a Destination"
              onChange={onChangeDestination}
              data-testid="select-destination-option">
              {optionValues.length > 0 &&
                optionValues.map(({ name, value }, ind) => (
                  <Option
                    value={value}
                    // eslint-disable-next-line react/no-array-index-key
                    key={ind}
                    className="referral-link-destination-option">
                    {name}
                  </Option>
                ))}
            </Select>
          </Form.Item>
          <Form.Item>
            <Button
              type="primary"
              htmlType="submit"
              className="button-common-class generate-link-button"
              data-testid="generate-link-button">
              Generate Link
            </Button>
          </Form.Item>
        </Form>
      </ReferralRowTemplate>
      {isModalOpen && (
        <ReferralLinkModal
          isModalOpen={isModalOpen}
          closeModal={() => setIsModalOpen(false)}
          holdClose={!!loading}
          isFailure={!(error === undefined)}
          generateReferralLink={
            data !== undefined ? data.generateReferralLink : undefined
          }
        />
      )}
    </>
  );
}

export default CreateReferral;
