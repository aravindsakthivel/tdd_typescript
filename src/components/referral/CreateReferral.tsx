import { FC, useState } from 'react';
import { Select, Button, Form } from 'antd';
import { ReferralRowTemplate } from './ReferralRowTemplate';
import { ReferralLinkModal } from './ReferralLinkModal';
import { SelectOptionInterface } from '../Interfaces/generalInterfaces';
import '../customClassStyle.css';

const { Option } = Select;

// the following interface was to test the select
interface CreateReferralTestInterface {
  destinationDefault?: 'mpl_pro' | 'mpl_qa' | 'mpl_dev';
}

const CreateReferral: FC<CreateReferralTestInterface> = ({ destinationDefault = '' }) => {
  let holdClose = false;
  let referralLink = 'www.mpl.com/mobileapp/livegame/mobileapp';
  let isFailure = false;
  const [isModalOpen, setIsModalOpen] = useState<boolean>(false);
  const [destination, setDestination] = useState<string>(destinationDefault);
  let optionValues: SelectOptionInterface[] = [
    {
      name: 'MPL PRO',
      value: 'mpl_pro',
    },
    {
      name: 'MPL QA',
      value: 'mpl_qa',
    },
    {
      name: 'MPL DEV',
      value: 'mpl_dev',
    },
  ];

  const onChangeDestination = (option: string) => {
    setDestination(option);
  };

  const onSubmit = (values: any) => {
    setIsModalOpen(!isModalOpen);
  };

  return (
    <>
      <ReferralRowTemplate title={'Create a custom link'}>
        <Form
          layout="inline"
          size={'large'}
          onFinish={onSubmit}
          data-testid="generate-link-form"
          initialValues={destinationDefault !== '' ? { destination: destinationDefault } : {}}
          className="referral-link-form"
        >
          <Form.Item
            name="destination"
            rules={[{ required: true, message: 'Please select a Destination' }]}
          >
            <Select
              style={{ width: 350, borderRadius: 10 }}
              placeholder="Select a Destination"
              onChange={onChangeDestination}
              data-testid="select-destination-option"
            >
              {optionValues.length > 0 &&
                optionValues.map(({ name, value }, ind) => (
                  <Option value={value} key={ind} className="referral-link-destination-option">
                    {name}
                  </Option>
                ))}
            </Select>
          </Form.Item>
          <Form.Item>
            <Button
              type="primary"
              htmlType="submit"
              className="button-common-class generate-link-button"
              data-testid="generate-link-button"
            >
              Generate Link
            </Button>
          </Form.Item>
        </Form>
      </ReferralRowTemplate>
      <ReferralLinkModal
        isModalOpen={isModalOpen}
        closeModal={() => setIsModalOpen(false)}
        holdClose={holdClose}
        destination={destination}
        referralLink={referralLink}
        isFailure={isFailure}
      />
    </>
  );
};

export { CreateReferral };
