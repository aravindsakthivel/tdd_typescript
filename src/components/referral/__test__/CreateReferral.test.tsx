import ReactDOM from 'react-dom';
import { render, cleanup, act, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import { mockConfig } from '../../mock';
import { CreateReferral } from '../CreateReferral';

afterEach(cleanup);
beforeEach(() => {
  mockConfig();
});

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<CreateReferral />, div);
});

it('component consit of select and button', () => {
  const { getByTestId } = render(<CreateReferral />);

  expect(getByTestId(/select-destination-option/i)).toBeInTheDocument();
  expect(getByTestId(/generate-link-button/i)).toBeInTheDocument();
});

describe('Form behavior', () => {
  it('component shows error when clicking button without selecting', async () => {
    const { getByTestId } = render(<CreateReferral />);

    await act(async () => {
      fireEvent.submit(getByTestId('generate-link-form'));
    });

    await screen.findByText('Please select a Destination');
  });

  it('validation passes form submits and modal appears', async () => {
    const { getByTestId } = render(<CreateReferral destinationDefault={'mpl_pro'} />);

    await act(async () => {
      fireEvent.submit(getByTestId('generate-link-form'));
    });

    await screen.findByText('New Link');
    expect(getByTestId('referral-modal-destination-title')).toHaveTextContent(
      'Destination: MPL PRO',
    );
    expect(screen.getByText('www.mpl.com/mobileapp/livegame/mobileapp')).toBeInTheDocument();
  });
});

it('matches snapshot without modal', () => {
  const tree = renderer.create(<CreateReferral />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('matches snapshot with modal', () => {
  const tree = renderer.create(<CreateReferral destinationDefault={'mpl_pro'} />).toJSON();
  expect(tree).toMatchSnapshot();
});
