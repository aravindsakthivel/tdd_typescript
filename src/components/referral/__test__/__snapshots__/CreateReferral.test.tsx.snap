// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot with modal 1`] = `
<div
  className="ant-row referral-row-template-class general-background-border false"
  style={Object {}}
>
  <div
    className="ant-col ant-col-24"
    style={Object {}}
  >
    <h5
      className="ant-typography"
      data-testid="referral-row-title"
      style={
        Object {
          "WebkitLineClamp": undefined,
        }
      }
    >
      Create a custom link
    </h5>
  </div>
  <div
    className="ant-col ant-col-23"
    data-testid="referral-row-children"
    style={Object {}}
  >
    <form
      className="ant-form ant-form-inline ant-form-large referral-link-form"
      data-testid="generate-link-form"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="ant-row ant-form-item"
        style={Object {}}
      >
        <div
          className="ant-col ant-form-item-control"
          style={Object {}}
        >
          <div
            className="ant-form-item-control-input"
          >
            <div
              className="ant-form-item-control-input-content"
            >
              <div
                className="ant-select ant-select-lg ant-select-single ant-select-show-arrow"
                data-testid="select-destination-option"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                style={
                  Object {
                    "borderRadius": 10,
                    "width": 350,
                  }
                }
              >
                <div
                  className="ant-select-selector"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                >
                  <span
                    className="ant-select-selection-search"
                  >
                    <input
                      aria-activedescendant="destination_list_0"
                      aria-autocomplete="list"
                      aria-controls="destination_list"
                      aria-haspopup="listbox"
                      aria-owns="destination_list"
                      autoComplete="off"
                      className="ant-select-selection-search-input"
                      id="destination"
                      onChange={[Function]}
                      onCompositionEnd={[Function]}
                      onCompositionStart={[Function]}
                      onKeyDown={[Function]}
                      onMouseDown={[Function]}
                      onPaste={[Function]}
                      readOnly={true}
                      role="combobox"
                      style={
                        Object {
                          "opacity": 0,
                        }
                      }
                      type="search"
                      unselectable="on"
                      value=""
                    />
                  </span>
                  <span
                    className="ant-select-selection-item"
                    title="MPL PRO"
                  >
                    MPL PRO
                  </span>
                </div>
                <span
                  aria-hidden={true}
                  className="ant-select-arrow"
                  onMouseDown={[Function]}
                  style={
                    Object {
                      "WebkitUserSelect": "none",
                      "userSelect": "none",
                    }
                  }
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    className="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-row ant-form-item"
        style={Object {}}
      >
        <div
          className="ant-col ant-form-item-control"
          style={Object {}}
        >
          <div
            className="ant-form-item-control-input"
          >
            <div
              className="ant-form-item-control-input-content"
            >
              <button
                className="ant-btn ant-btn-primary ant-btn-lg button-common-class generate-link-button"
                data-testid="generate-link-button"
                onClick={[Function]}
                type="submit"
              >
                <span>
                  Generate Link
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`matches snapshot without modal 1`] = `
<div
  className="ant-row referral-row-template-class general-background-border false"
  style={Object {}}
>
  <div
    className="ant-col ant-col-24"
    style={Object {}}
  >
    <h5
      className="ant-typography"
      data-testid="referral-row-title"
      style={
        Object {
          "WebkitLineClamp": undefined,
        }
      }
    >
      Create a custom link
    </h5>
  </div>
  <div
    className="ant-col ant-col-23"
    data-testid="referral-row-children"
    style={Object {}}
  >
    <form
      className="ant-form ant-form-inline ant-form-large referral-link-form"
      data-testid="generate-link-form"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="ant-row ant-form-item"
        style={Object {}}
      >
        <div
          className="ant-col ant-form-item-control"
          style={Object {}}
        >
          <div
            className="ant-form-item-control-input"
          >
            <div
              className="ant-form-item-control-input-content"
            >
              <div
                className="ant-select ant-select-lg ant-select-single ant-select-show-arrow"
                data-testid="select-destination-option"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                style={
                  Object {
                    "borderRadius": 10,
                    "width": 350,
                  }
                }
              >
                <div
                  className="ant-select-selector"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                >
                  <span
                    className="ant-select-selection-search"
                  >
                    <input
                      aria-activedescendant="destination_list_0"
                      aria-autocomplete="list"
                      aria-controls="destination_list"
                      aria-haspopup="listbox"
                      aria-owns="destination_list"
                      autoComplete="off"
                      className="ant-select-selection-search-input"
                      id="destination"
                      onChange={[Function]}
                      onCompositionEnd={[Function]}
                      onCompositionStart={[Function]}
                      onKeyDown={[Function]}
                      onMouseDown={[Function]}
                      onPaste={[Function]}
                      readOnly={true}
                      role="combobox"
                      style={
                        Object {
                          "opacity": 0,
                        }
                      }
                      type="search"
                      unselectable="on"
                      value=""
                    />
                  </span>
                  <span
                    className="ant-select-selection-placeholder"
                  >
                    Select a Destination
                  </span>
                </div>
                <span
                  aria-hidden={true}
                  className="ant-select-arrow"
                  onMouseDown={[Function]}
                  style={
                    Object {
                      "WebkitUserSelect": "none",
                      "userSelect": "none",
                    }
                  }
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    className="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-row ant-form-item"
        style={Object {}}
      >
        <div
          className="ant-col ant-form-item-control"
          style={Object {}}
        >
          <div
            className="ant-form-item-control-input"
          >
            <div
              className="ant-form-item-control-input-content"
            >
              <button
                className="ant-btn ant-btn-primary ant-btn-lg button-common-class generate-link-button"
                data-testid="generate-link-button"
                onClick={[Function]}
                type="submit"
              >
                <span>
                  Generate Link
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
